// @plasmicapp/loader/PlasmicLoader.jsx
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlasmicRootProvider = exports.Providers = exports.getComponentFromUrl = exports.getUnregisteredPageUrls = exports.getPageUrls = exports.componentsByPath = void 0;
function PlasmicLoader(props) {
    return null;
}
exports.default = PlasmicLoader;
PlasmicLoader.getPlumeType = function (props) {
    return undefined;
};
exports.componentsByPath = {};
function getPageUrls() {
    return [];
}
exports.getPageUrls = getPageUrls;
function getUnregisteredPageUrls() {
    return [];
}
exports.getUnregisteredPageUrls = getUnregisteredPageUrls;
function getComponentFromUrl(url) {
    throw new Error("This content should be overriden every time you run plasmic loader. If you are seeing this error, please report it as a bug!");
}
exports.getComponentFromUrl = getComponentFromUrl;
function Providers(props) {
    return null;
}
exports.Providers = Providers;
/**
 * @deprecated This package is no longer supported. We recommend updating to new version of the Headless API. https://docs.plasmic.app/learn/upgrading-to-loader-v2
 */
function PlasmicRootProvider(props) {
    return props.children;
}
exports.PlasmicRootProvider = PlasmicRootProvider;
