// @plasmicapp/loader/shared/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.crash = exports.error = exports.warn = exports.info = exports.cliError = exports.cliInfo = void 0;
const chalk_1 = __importDefault(require("chalk"));
const sentry_1 = require("./sentry");
const prefixes = {
    cliInfo: `${chalk_1.default.blueBright("info")} ${chalk_1.default.green("plasmic-cli")}`,
    cliError: `${chalk_1.default.red("error")} ${chalk_1.default.green("plasmic-cli")}`,
    info: `${chalk_1.default.blueBright("info")} ${chalk_1.default.green("plasmic-loader")}`,
    warn: `${chalk_1.default.yellow("warn")} ${chalk_1.default.green("plasmic-loader")}`,
    error: `${chalk_1.default.red("error")} ${chalk_1.default.green("plasmic-loader")}`,
};
function cliInfo(message) {
    console.log(`${prefixes.cliInfo} - ${message}`);
}
exports.cliInfo = cliInfo;
function cliError(message) {
    console.log(`${prefixes.cliError} - ${message}`);
}
exports.cliError = cliError;
function info(message) {
    console.log(`${prefixes.info} - ${message}`);
}
exports.info = info;
function warn(message) {
    console.warn(`${prefixes.warn} - ${message}`);
}
exports.warn = warn;
function error(message) {
    console.error(`${prefixes.error} - ${message}`);
}
exports.error = error;
function crash(message, err) {
    error(message);
    if (err) {
        console.error(err);
        sentry_1.captureException(err).then(() => process.exit(1));
        return;
    }
    process.exit(1);
}
exports.crash = crash;
